### ISSUE TITLE
[9-1] 研究 Isaac Lab 環境設計規範（gym.Env 介面）

### ISSUE THREAD

**Issue by GuanPersonalDev** [GuanPersonalDev/isaac-sim-digital-twin#120]
## 任務目標
在 RunPod container `nvcr.io/nvidia/isaac-lab:3.0.0-beta2` 內研究 Isaac Lab 環境設計規範，了解 gym.Env / ManagerBasedRLEnv 介面與環境註冊方式。

## 設計變更（2026-08-01）
訓練平台改為雲端（RunPod Community RTX 4090），使用 NVIDIA pre-built container，**本機不再需要安裝 Isaac Lab**。移除本機安裝步驟，改為在雲端容器內直接研究介面。

## 預估時數
1h

## 完成標準
- 確認 Isaac Lab 環境介面規範（gym.Env / ManagerBasedRLEnv）
- 了解 reset()、step()、observation_space、action_space 的實作要求
- 了解環境註冊方式（gym.register 或 Isaac Lab 慣例）

**Comment by GuanPersonalDev** [comment]
## 研究結論（2026-08-07）

### 研究方式與原訂的差異

Issue 原訂「在 RunPod container 內研究」，實際改用本機 `IsaacLab/` 原始碼直接查證
（`VERSION` = 3.0.0，與 pod 上同版），**未開 pod**。

好處是可完整瀏覽原始碼、對照行號；代價是**所有結論都來自靜態閱讀，未在執行環境實測**。
下方標記 ⚠️ 的三項尤其需要在 #121/#122 上機時實地確認。

---

## 1. 環境介面規範

Isaac Lab 的 env **本身就是向量化的**：一個 env 物件內含 `num_envs` 份平行子環境，
所有回傳值都是 GPU 上的 batched `torch.Tensor`。沒有 `SubprocVecEnv` 那一層。

```python
obs, extras = env.reset()
# obs["policy"]: (num_envs, obs_dim)

obs, reward, terminated, truncated, extras = env.step(action)
# action:     (num_envs, action_dim)
# reward:     (num_envs,)
# terminated: (num_envs,) bool  任務真的結束
# truncated:  (num_envs,) bool  時間到（episode_length_s）
```

與標準 gymnasium 的兩點差異：

- **`step()` 內部會自動 reset 已終止的環境**（`manager_based_rl_env.py:170`），
  不需自行判斷 done 再呼叫 reset。
- **`reset()` 是全體重置**；重置子集是內部的 `_reset_idx(env_ids)`，由框架呼叫。

`step()` 內的順序（`manager_based_rl_env.py:164-172`）：

```
process actions → physics ×decimation → render → 算 reward/termination
→ reset 已終止的環境 → 算 observation → 回傳
```

注意 **reward/termination 在 reset 之前算，observation 在 reset 之後算**，
所以 obs 拿到的是新 episode 的第一筆觀測。這是慣例，不是 bug。

子環境不是各自獨立的世界：只有一個 PhysX scene，N 份場景 prim 複製到
`/World/envs/env_.*`（`InteractiveScene.clone_environments()`），靠**空間位移**分開，
`scene.env_origins`（shape `(num_envs, 3)`）記錄每份的世界座標原點，
子環境之間靠 `filter_collisions()` 關掉碰撞。

> ⚠️ 對本專案的意涵：`core/` 算出來的是**桌台相對座標**，寫進模擬時必須加上
> `env_origins[env_id]` 才是世界座標。這是 #121 的已知接線點。

---

## 2. Workflow 選型：採用 Manager-based

`ManagerBasedRLEnv`（宣告式，填 manager cfg）與 `DirectRLEnv`（命令式，覆寫 `_get_*` 方法）
兩套並存。**決定採用 Manager-based**，理由：

1. **reward 分項自動進 TensorBoard**。`reward_manager.py:116-122` 會輸出
   `Episode_Reward/<term_name>`，每個 RewTerm 一條獨立曲線；
   `termination_manager.py:147` 同理有 `Episode_Termination/<key>`。
   撞球的 reward 必然要反覆調配比（散開／進袋／白球進袋／犯規），
   能一眼看出哪一項在主導，對 8/15 的 A-CP（#179）是實打實的除錯效率。
2. **Hydra CLI override**（`train.py:124` 的 `@hydra_task_config`）：
   `env.rewards.foul.weight=-0.5` 不改程式碼就能重跑，雲端掃參數省掉 commit-push-pull。
3. **`weight=0.0` 的純度量 term**（cartpole 有此用法）：不影響梯度、只借 logging 管道畫曲線，
   適合「進袋數」「散開熵」這類指標。
4. **IO descriptors 匯出只支援 Manager-based**（見第 5 節）。

原先曾傾向 Direct，理由是 `core/services/rl_observation_encoder.py` 是單一入口、
Manager-based 會逼你拆碎它。**該前提不成立**——Term 的 `func` 會拿到 `env`，
可以只放**一個** ObsTerm 包住現有編碼器，`core/` 一行都不用拆。

---

## 3. observation_space / action_space

兩種 workflow 都由 `_configure_gym_env_spaces()` 推導：

- Manager-based：從 ObservationManager 各 group 的形狀推出。
- Direct：cfg 上寫 `observation_space = 21`（int 會自動轉 `Box`）。

obs 是 **dict**，至少有 `"policy"` group；非對稱 actor-critic 會多一個 `"critic"`。
`env.observation_space` 是 single-env 的 space。

> ⚠️ 連帶結論：`training/configs/ppo_billiard.yaml` 的 `obs_dim` / `action_dim`
> 兩欄應刪除。Isaac Lab 完全從 env cfg 推導，yaml 那兩欄不會被讀，
> 只會變成會腐爛的第二份規格。（`training/README.md` 的〈尚未完成〉已記此事，此處確認判斷正確。）

---

## 4. 環境註冊方式

標準寫法（`isaaclab_tasks/manager_based/classic/cartpole/__init__.py`）：

```python
gym.register(
    id="Isaac-Cartpole-v0",
    entry_point="isaaclab.envs:ManagerBasedRLEnv",   # 環境類別
    disable_env_checker=True,
    kwargs={
        "env_cfg_entry_point": f"{__name__}.cartpole_env_cfg:CartpoleEnvCfg",   # config 類別
        "rsl_rl_cfg_entry_point": f"{agents.__name__}.rsl_rl_ppo_cfg:CartpolePPORunnerCfg",
    },
)
```

`entry_point`（環境類別）與 `env_cfg_entry_point`（config 類別）**分開**是 Isaac Lab 慣例，
Hydra 靠後者做 `env.xxx=yyy` 的 CLI override。

**外部 repo 的坑**：`train.py:41` 只 `import isaaclab_tasks`，不會知道
`/workspace/isaac-sim-digital-twin` 裡有我們的 `gym.register`，
沒被 import 就直接 `NameNotFound`。兩條解法：

- `--external_callback my_pkg.registration:register_envs`（`train.py:85, 97-99` 官方預留給外部 repo）
- 把 task 做成可安裝 package，`pip install -e` 後 import 即註冊（`./isaaclab.sh -n` 產生的
  External 專案就是這個形式，官方標為 recommended）

二擇一，在 #121 決定並記錄。

---

## 5. 匯出有三種，不要混為一談

### A. Policy 匯出（TorchScript / ONNX）— #226/#227 要的

`play.py` 每跑一次自動產生，落點 `<checkpoint 目錄>/exported/policy.pt` 與 `policy.onnx`。
與 workflow 無關（匯出器只認 `policy.actor` 這個 nn.Module）。

> ⚠️ **`training/README.md` 需修正。** `play_rsl_rl.py:174-196` 有版本分歧：
> rsl_rl >= 4.0.0 走 `runner.export_policy_to_jit()` / `export_policy_to_onnx()`，
> 低於才走 isaaclab 自己的 exporter。而 `train.py:59` 寫死
> `RSL_RL_VERSION = "5.0.1"`（低於就 `exit(1)`），所以**一定走前者**。
>
> README 記的「ONNX `dynamic_axes={}`、batch size 固定為 1」出自 isaaclab 的
> `_OnnxPolicyExporter`（`exporter.py:188, 215`），**該路徑在此環境不會執行**。
> 同理「normalizer 已打包進模型」我只確認了 isaaclab 版本
> （`exporter.py:101`：`self.actor(self.normalizer(x))`）；rsl_rl 5.x 的
> `export_policy_to_jit` 原始碼不在 IsaacLab repo 內，未查證。
>
> 這兩點直接影響 #227 本機回放（要不要自己正規化、batch 能否 > 1）。
> **在 pod 上實際匯出一次後打開檔案確認，不要照舊筆記走。**

### B. IO descriptors 匯出 — Manager-based 專屬

`--export_io_descriptors` → `<log_dir>/io_descriptors/IO_descriptors.yaml`
（`manager_based_env.py:313-339`），內容四塊：observations / actions / articulations / scene。
`train.py:180-186` 明確判斷：非 `ManagerBasedRLEnvCfg` 只印 warning 跳過。

價值：#227 是本機用 Isaac Sim 內建 PyTorch 載 TorchScript、餵原始觀測。
**餵錯欄位順序不會報錯，只會行為詭異。** 這份 yaml 就是機器可讀的契約
（shape / dtype / 單位 / 軸向 / joint 名稱），可在本機寫成斷言。

代價：自訂 observation 函式要掛 `@generic_io_descriptor(...)`（`io_descriptors.py:118`）。
沒掛的話 `get_IO_descriptors` 是包在 try/except 裡的，只 print 一行錯誤並
**靜默跳過該 term**——yaml 照樣產出但少一塊，這個失敗模式要知道。

### C. cfg dump — 兩種 workflow 都有

`train.py:241-242` 每次訓練把完整設定寫進 `<log_dir>/params/env.yaml` 與 `agent.yaml`。
配合「評估場景必須完全固定」的要求，回放時可 diff 兩個 checkpoint 的 env.yaml
確認除了 policy 以外真的沒變。

---

## 6. Headless 執行（RunPod）

**`--headless` 在 3.0 已 deprecated**（`app_launcher.py:488-495`）：
新模型是 `--visualizer` / `--viz`，接 CSV（`kit,newton,rerun,viser`），**不給就是 headless**。

**入口也換了。** `training/scripts/run_train.sh` 預留的
`scripts/reinforcement_learning/rsl_rl/train.py` 開頭第 10-16 行就是 `DeprecationWarning`。
3.0 統一入口（`isaaclab/cli/__init__.py:49` 確認 `train` 子指令存在）：

```bash
/workspace/IsaacLab/isaaclab.sh train --rl_library rsl_rl --task Isaac-Billiard-v0 --num_envs 1024
```

> ⚠️ **checkpoint 落點是 CWD 相對路徑。** `train.py:167-168`：
> `os.path.abspath(os.path.join("logs", "rsl_rl", experiment_name))`。
> `run_train.sh` 的 `OUTPUT_DIR` / `TRAIN_OUTPUT` 只被 `mkdir -p` 建出來，
> 沒有任何參數把它傳進去——**也沒有這種參數存在**。
> 實際落點是 `$PWD/logs/rsl_rl/<experiment>/<timestamp>/`。
> 若在 container disk 目錄啟動訓練，**停機就全沒了**，正是 README 一直在防的情況。
> 要靠 `cd` 到 `/workspace` 底下控制，或把 `logs` symlink 到 volume。

### headless 下看得到什麼

| 管道 | 說明 |
|---|---|
| stdout | rsl_rl 每 iteration 印 mean reward / episode length |
| TensorBoard | port 6006，`--host 0.0.0.0` 不可省 |
| checkpoint | `model_<iter>.pt`；`play.py` 跑一次順便匯出 TorchScript |
| `--video` | 自動設 `enable_cameras=True`（`train.py:90-91`），mp4 落在 `<log_dir>/videos/train/`。**Block 12 Demo 素材的第二條路**，代價是 RTX 渲染拖慢訓練，`--video_interval` 要開稀疏 |
| viser | `--viz viser`，預設 port **8080**、`bind_address="0.0.0.0"`，docstring 明寫供 remote access。**HTTP/WebSocket，RunPod 的 TCP proxy 轉得過去** |

> viser 與 README 記的「WebRTC 需 UDP、RunPod 只轉 TCP」不衝突——那條結論仍成立，
> viser 是另一條路。但要誠實看待：viser 畫的是幾何體，**不是 RTX 畫面**，
> Demo 說服力不如本機回放；1024 環境推到瀏覽器會很慘。
> 正確定位是**少量環境的除錯工具**（`--num_envs 4`），
> 解決「機器人穿模了嗎」「球生在桌外嗎」這類目前只能印 tensor 猜的問題。

`manager_based_rl_env.py:181-183` 補充：Newton / Rerun / Viser 的更新**獨立於 Kit app loop**，
`render_enabled=False` 時照常更新——這是它們比 Kit GUI 適合遠端的原因。

---

## 7. 對其他檔案／Issue 的連帶影響（待處理）

| 項目 | 位置 | 動作 |
|---|---|---|
| 訓練入口已 deprecated | `training/scripts/run_train.sh` | 改為 `isaaclab.sh train --rl_library rsl_rl` |
| `TRAIN_OUTPUT` 不生效 | `training/scripts/run_train.sh` | 改成靠 `cd` 控制，或拿掉以免誤導 |
| ONNX batch=1 / normalizer 打包 | `training/README.md` | 該敘述來自不會執行的程式路徑，需上機實測後改寫 |
| `obs_dim` / `action_dim` | `training/configs/ppo_billiard.yaml` | 刪除（Isaac Lab 從 env 推導） |
| 外部 repo 註冊機制 | #121 | `--external_callback` vs 可安裝 package，二擇一 |
| Workflow 決定 | #121 | Manager-based，已寫入該 issue 的實作拆分 |

**Comment by GuanPersonalDev** [comment]
完成標準三項全部達成，研究結論已寫入上方 memo。實作決策（Manager-based workflow）與拆分已寫進 #121。
